File: backend/app/routers/medicines.py
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_medicine_db
from app.schemas.medicine import (
    MedicineAlternativesResponse,
    MedicineListItem,
    MedicineResponse,
    MedicineSearchResponse,
)
from app.services import medicine_service

router = APIRouter(prefix="/medicines", tags=["medicines"])


@router.get("/search", response_model=MedicineSearchResponse)
async def search_medicines(
    q: Optional[str] = Query(None, description="Search by brand name or manufacturer"),
    component_id: Optional[UUID] = Query(None, description="Filter by component ID"),
    therapeutic_class: Optional[str] = Query(None, description="Filter by therapeutic class"),
    include_discontinued: bool = Query(False, description="Include discontinued medicines"),
    limit: int = Query(50, ge=1, le=100, description="Results per page"),
    page: int = Query(1, ge=1, description="Page number"),
    db: AsyncSession = Depends(get_medicine_db),
):
    """
    Search medicines with filters
    - Search by brand name or manufacturer
    - Filter by component (get all medicines containing specific component)
    - Filter by therapeutic class
    - Pagination support
    """
    offset = (page - 1) * limit

    medicines, total = await medicine_service.search_medicines(
        db=db,
        search=q,
        component_id=component_id,
        therapeutic_class=therapeutic_class,
        include_discontinued=include_discontinued,
        limit=limit,
        offset=offset,
    )

    # Convert to list items with component info
    medicine_items = []
    for med in medicines:
        components = [
            {
                "component_id": mc.component_id,
                "component_name": mc.component.name,
                "strength": mc.strength,
                "unit": mc.unit,
                "sequence": mc.sequence,
            }
            for mc in sorted(med.medicine_components, key=lambda x: x.sequence)
        ]

        medicine_items.append({
            "id": med.id,
            "brand_name": med.brand_name,
            "manufacturer": med.manufacturer,
            "dosage_form": med.dosage_form,
            "strength": med.strength,
            "mrp": med.mrp,
            "is_discontinued": med.is_discontinued,
            "components": components,
        })

    pages = (total + limit - 1) // limit if total > 0 else 0

    return {
        "medicines": medicine_items,
        "total": total,
        "page": page,
        "pages": pages,
    }


@router.get("/{medicine_id}", response_model=MedicineResponse)
async def get_medicine(
    medicine_id: UUID,
    db: AsyncSession = Depends(get_medicine_db),
):
    """
    Get medicine details by ID
    - Full medicine information
    - All components with strengths
    - Interactions, side effects, uses
    - Alternatives
    """
    medicine = await medicine_service.get_medicine_by_id(db, medicine_id)
    if not medicine:
        raise HTTPException(status_code=404, detail="Medicine not found")

    # Build component response
    components = [
        {
            "component_id": mc.component_id,
            "component_name": mc.component.name,
            "strength": mc.strength,
            "unit": mc.unit,
            "sequence": mc.sequence,
        }
        for mc in sorted(medicine.medicine_components, key=lambda x: x.sequence)
    ]

    # Compute salt composition
    salt_composition = " + ".join([
        f"{mc.component.name} ({mc.strength}{mc.unit})"
        for mc in sorted(medicine.medicine_components, key=lambda x: x.sequence)
    ])

    return {
        "id": medicine.id,
        "brand_name": medicine.brand_name,
        "manufacturer": medicine.manufacturer,
        "dosage_form": medicine.dosage_form,
        "strength": medicine.strength,
        "pack_size": medicine.pack_size,
        "therapeutic_class": medicine.therapeutic_class,
        "schedule": medicine.schedule,
        "mrp": medicine.mrp,
        "is_discontinued": medicine.is_discontinued,
        "habit_forming": medicine.habit_forming,
        "alternatives": medicine.alternatives,
        "interactions": medicine.interactions,
        "components": components,
        "salt_composition": salt_composition,
        "created_at": medicine.created_at,
        "updated_at": medicine.updated_at,
    }


@router.get("/{medicine_id}/alternatives", response_model=MedicineAlternativesResponse)
async def get_medicine_alternatives(
    medicine_id: UUID,
    db: AsyncSession = Depends(get_medicine_db),
):
    """
    Get alternative/substitute medicines for a given medicine
    """
    result = await medicine_service.get_medicine_alternatives(db, medicine_id)
    if not result:
        raise HTTPException(status_code=404, detail="Medicine not found")

    return result

File: frontend/src/lib/api/medicines.ts
/**
 * Medicine API Client
 * Connects to backend medicine endpoints
 */

import api from '@/lib/api';

export interface MedicineComponent {
  component_id: string;
  component_name: string;
  strength: number | string;  // Can be string from API (Decimal type)
  unit: string;
  sequence: number;
}

export interface Medicine {
  id: string;
  brand_name: string;
  manufacturer: string | null;
  dosage_form: string | null;
  strength: string | null;
  pack_size: string | null;
  therapeutic_class: string | null;
  schedule: string | null;
  mrp: number | string | null;  // Can be string from API (Decimal type)
  is_discontinued: boolean;
  habit_forming: boolean;
  components: MedicineComponent[];
  alternatives?: any;
  interactions?: {
    side_effects?: string[];
    uses?: string[];
    chemical_class?: string;
    action_class?: string;
  };
  salt_composition?: string;
  created_at?: string;
  updated_at?: string;
}

export interface MedicineSearchParams {
  search?: string;
  component_id?: string;
  therapeutic_class?: string;
  include_discontinued?: boolean;
  limit?: number;
  page?: number;
}

export interface MedicineSearchResponse {
  medicines: Medicine[];
  total: number;
  page: number;
  pages: number;
}

export interface Component {
  id: string;
  name: string;
  common_names: string | null;
  category: string | null;
  description: string | null;
  medicine_count?: number;
  created_at: string;
  updated_at: string;
}

export interface ComponentSearchResponse {
  components: Component[];
  total: number;
}

/**
 * Search medicines with filters
 */
export async function searchMedicines(
  params: MedicineSearchParams = {}
): Promise<MedicineSearchResponse> {
  const queryParams = new URLSearchParams();

  if (params.search) queryParams.append('q', params.search);
  if (params.component_id) queryParams.append('component_id', params.component_id);
  if (params.therapeutic_class) queryParams.append('therapeutic_class', params.therapeutic_class);
  if (params.include_discontinued !== undefined) {
    queryParams.append('include_discontinued', String(params.include_discontinued));
  }
  if (params.limit) queryParams.append('limit', String(params.limit));
  if (params.page) queryParams.append('page', String(params.page));

  return (await api.get(`/api/v1/medicines/search?${queryParams}`)).data;
}

/**
 * Get medicine by ID
 */
export async function getMedicine(id: string): Promise<Medicine> {
  return (await api.get(`/api/v1/medicines/${id}`)).data;
}

/**
 * Get medicine alternatives
 */
export async function getMedicineAlternatives(id: string): Promise<any> {
  return (await api.get(`/api/v1/medicines/${id}/alternatives`)).data;
}

/**
 * Search components (Admin)
 */
export async function searchComponents(
  search?: string,
  limit: number = 50,
  offset: number = 0
): Promise<ComponentSearchResponse> {
  const queryParams = new URLSearchParams();

  if (search) queryParams.append('search', search);
  queryParams.append('limit', String(limit));
  queryParams.append('offset', String(offset));

  return (await api.get(`/api/v1/admin/components?${queryParams}`)).data;
}

/**
 * Create medicine (Admin)
 */
export async function createMedicine(data: {
  brand_name: string;
  manufacturer?: string;
  components: Array<{
    component_id: string;
    strength: number;
    unit: string;
  }>;
  dosage_form?: string;
  pack_size?: string;
  therapeutic_class?: string;
  mrp?: number;
  is_discontinued?: boolean;
  habit_forming?: boolean;
  alternatives?: any;
  interactions?: any;
}): Promise<Medicine> {
  return (await api.post('/api/v1/admin/medicines', data)).data;
}

/**
 * Update medicine (Admin)
 */
export async function updateMedicine(
  id: string,
  data: Partial<{
    brand_name: string;
    manufacturer: string;
    components: Array<{
      component_id: string;
      strength: number;
      unit: string;
    }>;
    dosage_form: string;
    pack_size: string;
    therapeutic_class: string;
    mrp: number;
    is_discontinued: boolean;
    habit_forming: boolean;
    alternatives: any;
    interactions: any;
  }>
): Promise<Medicine> {
  return (await api.put(`/api/v1/admin/medicines/${id}`, data)).data;
}

/**
 * Delete medicine (Admin)
 */
export async function deleteMedicine(id: string): Promise<void> {
  await api.delete(`/api/v1/admin/medicines/${id}`);
}

/**
 * Get database statistics
 */
export async function getMedicineStats(): Promise<{
  total: number;
  active: number;
  discontinued: number;
  components: number;
}> {
  // Get total and discontinued from main search
  const [allResult, discontinuedResult, componentsResult] = await Promise.all([
    searchMedicines({ include_discontinued: true, limit: 1 }),
    searchMedicines({ include_discontinued: true, limit: 1 }).then(async () => {
      // Count discontinued separately
      const resp = await api.get('/api/v1/medicines/search?include_discontinued=true&limit=1000');
      return resp.data.medicines.filter((m: Medicine) => m.is_discontinued).length;
    }),
    searchComponents('', 1, 0),
  ]);

  const total = allResult.total;
  const discontinued = discontinuedResult;
  const active = total - discontinued;
  const components = componentsResult.total;

  return {
    total,
    active,
    discontinued,
    components,
  };
}
